### PR TITLE
feat: group video UI grid

### PR DESCRIPTION
**Pull request by atomrc**
## Type of change

Adds the `group-video-grid` component. Able to display all the participants streams in a standalone grid

### CODE REVIEW

File: app/script/components/groupVideoGrid.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+function arrayDiff(a, b) {
+  return b.filter(i => a.indexOf(i) === -1);
+}
+
+ko.components.register('group-video-grid', {
+  template: `
+    <div class="video-grid" data-bind="foreach: { data: grid, as: 'streamId' }">
+      <!-- ko if: streamId !== 0 -->
+      <video class="video-grid__element" autoplay data-bind="css: $parent.getClassNameForVideo($index()), sourceStream: $parent.getParticipantStream(streamId), muteMediaElement: $parent.getParticipantStream(streamId)">
+      </video>
```
Comment (lipis) [inline app/script/components/groupVideoGrid.js]: indent the `<video>` tag

File: app/style/components/group-video-grid.less [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  height: 100%;
+}
+.video-grid__element--full-size {
+  width: 100%;
+  height: 100%;
+}
+.video-grid__element--empty {
+  overflow: hidden;
+  width: 0;
+  height: 0;
```
Comment (lipis) [inline app/style/components/group-video-grid.less]:
```less
.video-grid {
  position: relative;
  width: 100%;
  height: 100vh;
  
  &__thumbnail {
    position: absolute;
    top: 1em;
    ...
  }

  &__element {
    position: absolute;
    width: 50%;
  }
  ...
}
 
```

File: app/style/components/group-video-grid.less [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  right: 1em;
+  width: 7em;
+  height: 6em;
+  border: solid 2px white;
+}
+.video-grid__element {
+  position: absolute;
+  width: 50%;
+  height: 50%;
+  border: solid 2px white;
```
Comment (lipis) [inline app/style/components/group-video-grid.less]: `#fff` (we are never using the named colors)

File: app/style/components/group-video-grid.less [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  height: 100vh;
+}
+
+.video-grid__thumbnail {
+  position: absolute;
+  top: 1em;
+  right: 1em;
+  width: 7em;
+  height: 6em;
+  border: solid 2px white;
```
Comment (lipis) [inline app/style/components/group-video-grid.less]: `#fff`

File: app/style/components/group-video-grid.less (hunk 2 above)
Comment (atomrc) [reply app/style/components/group-video-grid.less]: Oh, nice trick. Didn't know this one :+1:

File: app/script/components/groupVideoGrid.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      extraClass += ' video-grid__element--full-size';
+    } else if (!hasVerticalNeighbor) {
+      extraClass += ' video-grid__element--full-height';
+    }
+    return `${baseClass} ${extraClass}`;
+  }
+};
+
+function arrayDiff(a, b) {
+  return b.filter(i => a.indexOf(i) === -1);
```
Comment (lipis) [inline app/script/components/groupVideoGrid.js]: Didn't  we have a rule for minimum length for the variables?

Comment (atomrc) [reply app/script/components/groupVideoGrid.js]: It seems to be `1` :laughing:

File: app/style/components/group-video-grid.less (hunk 2 above)
Comment (lipis) [reply app/style/components/group-video-grid.less]: Works with SCSS as well.. :)

Also have an empty line between every rule.. apparently Prettier is not doing it :(

File: app/script/components/groupVideoGrid.js (hunk 5 above)
Comment (lipis) [reply app/script/components/groupVideoGrid.js]: LOL! :) Let me check who edited it.. :P

File: app/style/components/group-video-grid.less [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  }
+  &__element--full-size {
+    width: 100%;
+    height: 100%;
+  }
+  &__element--empty {
+    overflow: hidden;
+    width: 0;
+    height: 0;
+  }
```
Comment (lipis) [inline app/style/components/group-video-grid.less]: In theory we could also do some nesting with the `&__element*` as well.. but I'm not sure if @AndyLnd will agree :)

Comment (AndyLnd) [reply app/style/components/group-video-grid.less]: I agree, that's why we use a css preprocessor :)
also i do it the same way.

Comment (lipis) [reply app/style/components/group-video-grid.less]: Cool! Letz do diz then!

File: app/script/components/groupVideoGrid.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  getClassNameForVideo(index) {
+    const baseClass = `video-grid__element${index}`;
+    const grid = this.grid();
+    let extraClass = '';
+    if (grid[index] === 0) {
+      return `${baseClass} video-grid__element--empty`;
+    }
+    const isAlone = grid.reduce((alone, value, i) => (i !== index && value !== 0 ? false : alone), true);
```
Comment (AndyLnd) [inline app/script/components/groupVideoGrid.js]: `grid.every((value, i) => i === index || value === 0)`

File: app/script/components/groupVideoGrid.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+      extraClass += ' video-grid__element--full-size';
+    } else if (!hasVerticalNeighbor) {
+      extraClass += ' video-grid__element--full-height';
+    }
+    return `${baseClass} ${extraClass}`;
+  }
+};
+
+function arrayDiff(array1, array2) {
+  return array2.filter(element => array1.indexOf(element) === -1);
```
Comment (AndyLnd) [inline app/script/components/groupVideoGrid.js]: `array.includes(element)`

**Pull request merged**